src/vm/castcache.h

```diff
@@ -0,0 +1,356 @@
+        //
+        // Readers will read the version twice before and after retrieving the entry. 
+        // To have a usable entry both reads must yield the same even version.
+        // 
+        DWORD               version;
+        TADDR               source;
+        // pointers have unused lower bits due to alignment, we use one for the result
+        TADDR               targetAndResult;
+
+        FORCEINLINE TADDR Source()
+        {
+            return source;
+        }
+
+        FORCEINLINE TADDR Target()
+        {
+            return targetAndResult & ~(TADDR)1;
+        }
+
+        FORCEINLINE BOOL Result()
+        {
+            return targetAndResult & 1;
+        };
+
+        FORCEINLINE void SetEntry(TADDR source, TADDR target, BOOL result)
+        {
+            this->source = source;
+            this->targetAndResult = target | (result & 1);
+        }
+    };
+
+public:
+
+    FORCEINLINE static void TryAddToCache(TypeHandle source, TypeHandle target, BOOL result)
+    {
+        CONTRACTL
+        {
+            THROWS;
+            GC_TRIGGERS;
+            MODE_COOPERATIVE;
+        }
+        CONTRACTL_END;
+
+        // fully loaded types cannot be "undone" and thus castability can be cached.
+        // do not cache if any of the types is not fully loaded.
+        if (!source.IsFullyLoaded() || !target.IsFullyLoaded())
+            return;
+
+        // we should not be caching T --> Nullable<T>. result is contextual.
+        // unsubsttituted generic T is ok though. there is an agreement on that.
+        _ASSERTE(source.IsTypeDesc() || !Nullable::IsNullableForType(target, source.AsMethodTable()));
+
+        TryAddToCache(source.AsTAddr(), target.AsTAddr(), result);
+    }
+
+    FORCEINLINE static void TryAddToCache(MethodTable* pSourceMT, TypeHandle target, BOOL result)
+    {
+        CONTRACTL
+        {
+            THROWS;
+            GC_TRIGGERS;
+            MODE_COOPERATIVE;
+        }
+        CONTRACTL_END;
+
+        // fully loaded types cannot be "undone" and thus castability can be cached.
+        // do not cache if any of the types is not fully loaded.
+        if (!pSourceMT->IsFullyLoaded() || !target.IsFullyLoaded())
+            return;
+
+        // we should not be caching T --> Nullable<T>. result is contextual.
+        _ASSERTE(!Nullable::IsNullableForType(target, pSourceMT));
+
+        TryAddToCache((TADDR)pSourceMT, target.AsTAddr(), result);
+    }
+
+    FORCEINLINE static TypeHandle::CastResult TryGetFromCache(TypeHandle source, TypeHandle target)
+    {
+        CONTRACTL
+        {
+            NOTHROW;
+            GC_NOTRIGGER;
+            MODE_COOPERATIVE;
+        }
+        CONTRACTL_END;
+
+        return TryGetFromCache(source.AsTAddr(), target.AsTAddr());
+    }
+
+    FORCEINLINE static TypeHandle::CastResult TryGetFromCacheAny(TypeHandle source, TypeHandle target)
```

Comment:
This looks unused.

Reply:
Yes unused. I think all callers were from NoGC code paths and gone now.